diesel/src/connection/statement_cache/strategy.rs

```diff
@@ -0,0 +1,392 @@
+            name -> Text,
+        }
+    }
+
+    pub fn connection() -> PgConnection {
+        let mut conn = PgConnection::establish(&pg_database_url()).unwrap();
+        conn.set_instrumentation(RecordCacheEvents::default());
+        conn
+    }
+
+    #[test]
+    fn prepared_statements_are_cached() {
+        let connection = &mut connection();
+
+        let query = crate::select(1.into_sql::<Integer>());
+
+        assert_eq!(Ok(1), query.get_result(connection));
+        assert_eq!(1, count_cache_calls(connection));
+        assert_eq!(Ok(1), query.get_result(connection));
+        assert_eq!(1, count_cache_calls(connection));
+    }
+
+    #[test]
+    fn queries_with_identical_sql_but_different_types_are_cached_separately() {
+        let connection = &mut connection();
+
+        let query = crate::select(1.into_sql::<Integer>());
+        let query2 = crate::select("hi".into_sql::<VarChar>());
+
+        assert_eq!(Ok(1), query.get_result(connection));
+        assert_eq!(1, count_cache_calls(connection));
+        assert_eq!(Ok("hi".to_string()), query2.get_result(connection));
+        assert_eq!(2, count_cache_calls(connection));
+    }
+
+    #[test]
+    fn queries_with_identical_types_and_sql_but_different_bind_types_are_cached_separately() {
+        let connection = &mut connection();
+
+        let query = crate::select(1.into_sql::<Integer>()).into_boxed::<Pg>();
+        let query2 = crate::select("hi".into_sql::<VarChar>()).into_boxed::<Pg>();
+
+        assert_eq!(Ok(1), query.get_result(connection));
+        assert_eq!(1, count_cache_calls(connection));
+        assert_eq!(Ok("hi".to_string()), query2.get_result(connection));
+        assert_eq!(2, count_cache_calls(connection));
+    }
+
+    define_sql_function!(fn lower(x: VarChar) -> VarChar);
+
+    #[test]
+    fn queries_with_identical_types_and_binds_but_different_sql_are_cached_separately() {
+        let connection = &mut connection();
+
+        let hi = "HI".into_sql::<VarChar>();
+        let query = crate::select(hi).into_boxed::<Pg>();
+        let query2 = crate::select(lower(hi)).into_boxed::<Pg>();
+
+        assert_eq!(Ok("HI".to_string()), query.get_result(connection));
+        assert_eq!(1, count_cache_calls(connection));
+        assert_eq!(Ok("hi".to_string()), query2.get_result(connection));
+        assert_eq!(2, count_cache_calls(connection));
+    }
+
+    #[test]
+    fn queries_with_sql_literal_nodes_are_not_cached() {
+        let connection = &mut connection();
+        let query = crate::select(sql::<Integer>("1"));
+
+        assert_eq!(Ok(1), query.get_result(connection));
+        assert_eq!(0, count_cache_calls(connection));
+    }
+
+    #[test]
+    fn inserts_from_select_are_cached() {
+        let connection = &mut connection();
+        connection.begin_test_transaction().unwrap();
+
+        crate::sql_query(
+            "CREATE TABLE IF NOT EXISTS users(id INTEGER PRIMARY KEY, name TEXT NOT NULL);",
```

Comment:
Can we use `CREATE TEMPORARY TABLE` statements here instead? That prevents possible conflicts.